regparser/notice/xml.py

```diff
@@ -41,6 +42,100 @@ def _set_date_attr(self, date_type, value):
             value = value.isoformat()
         dates_tag.attrib["eregs-{}-date".format(date_type)] = value
 
+    def derive_agencies(self, agencies=[]):
+        """
+        SIDE EFFECTS: this operates on the XML of the NoticeXML itself as well
+        as returning some information.
+
+        Adds elements to the NoticeXML to reflect information about the
+        agencies connected to to notice.
+
+        Looks for that information in a list of dicts passed in as
+        ``agencies``, then adds it to the beginning of the XML as a set of
+        elements that will look something like this::
+
+
+            <EREGS_AGENCIES>
+                <EREGS_AGENCY eregs-agency-name="x" eregs-agency-id="00"
+                 eregs-agency-raw-name="X">
+                    <EREGS_SUBAGENCY eregs-agency-name="y" eregs-agency-id="01"
+                    eregs-agency-raw-name="Y"></EREGS_SUBAGENCY>
+                </EREGS_AGENCY>
+            </EREGS_AGENCIES>
+
+        :arg list agencies: dict objects containing agency information,
+                            including ``id``, ``parent_id``, ``name``, and
+                            ``raw_name``.
+
+        :rtype: dict
+        :returns:   A dict of ``id``: ``defaultdict``, where the id is
+                    the id of the agency, and the ``defaultdicts`` are nested
+                    to reflect their parent/child relationships.
+        """
+
+        if not agencies:
+            # The FR Notice XML doesn't tend to have all the metadata we need
+            # contained within it, so don't try to parse that, just log an
+            # error.
+            logging.warn("Preprocessing notice: no agency metadata.")
+            return {}
+
+        # We need turn turn the references to parent_ids into a tree of dicts
+        # that contain subagencies in children fields:
+        agency_map = {agency["id"]:
+                      defaultdict(list, agency) for agency in agencies}
```

Comment:
I wonder if the `defaultdict` here isn't overkill. What about something like:
```python
for agency in agencies:
    agency["children"] = []
agency_map = {agency["id"]: agency for agency in agencies}
# or, the less readable
agency_map = {agency["id"]: dict(children=[], **agency) for agency in agencies}
```
Both seems a bit more explicit in what you're trying to accomplish (if I'm reading this right).